### PR TITLE
[BEAM-6150] Superinterface for SerializableFunction allowing declared exceptions

### PR DESCRIPTION
**Pull request by jklukas**
Also provides an equivalent superclass for SimpleFunction.

See https://issues.apache.org/jira/browse/BEAM-6150

The intention here is to be both binary and recompile compatible with code using previous version of the Beam Java SDK, so some careful review is needed to ensure that's the case. See the Java specification's chapter on [binary compatibility](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html).

We originally discussed deprecating SerializableFunction and SimpleFunction in favor of the new classes, but there appear to be two fairly separate use cases for SerializableFunction. It's either defining user code that will be executed in a DoFn, in which case I think we always want to prefer the new interface that allows declared exceptions. But it's also used where the code is to be executed as part of pipeline construction, in which case it may be reasonable to want to restrict checked exceptions. In any case, deprecating SerializableFunction and SimpleFunction can be discussed further in the future.

I've attempted to provide significant context in the docstrings to explain the differences between these various classes and interfaces, and to explain the naming. The new interface and class are now used in `MapElements`, `FlatMapElements` and `Filter` as part of this PR. There are many other places in the SDK where we could alter methods to accept these less restrictive types, but those can be altered as needed in future changes.

I'm very open to feedback about suggestions for alternate names for the new `ProcessFunction` and `InferableFunction`.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by jklukas** [comment]
cc @reuvenlax and @rmannibucau as potential reviewers. If you have suggestions for how to validate that this will indeed be binary compatible, let me know.

**Comment by jklukas** [comment]
retest this please

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java [hunk 1]
```diff
@@ -166,12 +231,31 @@ public Integer apply(KV<Integer, String> input) {
   }
 
   /**
-   * Basic test of {@link MapElements} with a {@link SerializableFunction}. This style is generally
-   * discouraged in Java 7, in favor of {@link SimpleFunction}.
+   * Test of {@link MapElements} coder propagation with a parametric {@link InferableFunction} where
+   * the type variable occurs nested within other concrete type constructors.
```
Comment (jklukas) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: This looks out of place in the diff. The removed lines are the docstring for `testMapBasicSerializableFunction` which has been renamed to `testMapBasicProcessFunction` and appears as the next function below.

File: website/src/contribute/ptransform-style-guide.md [hunk 2]
```diff
@@ -395,8 +395,8 @@ If the transform has an aspect of behavior to be customized by a user's code, ma
...
+* If possible, just use PTransform composition as an extensibility device - i.e. if the same effect can be achieved by the user applying the transform in their pipeline and composing it with another `PTransform`, then the transform itself should not be extensible. E.g., a transform that writes JSON objects to a third-party system should take a `PCollection<JsonObject>` (assuming it is possible to provide a `Coder` for `JsonObject`), rather than taking a generic `PCollection<T>` and a `ProcessFunction<T, JsonObject>` (anti-example that should be fixed: `TextIO`).
+* If extensibility by user code is necessary inside the transform, pass the user code as a `ProcessFunction` or define your own serializable function-like type (ideally single-method, for interoperability with Java 8 lambdas). Because Java erases the types of lambdas, you should be sure to have adequate type information even if a raw-type `ProcessFunction` is provided by the user. See `MapElements` and `FlatMapElements` for examples of how to use `ProcessFunction` and `InferableFunction` in tandem to provide good support for both lambdas and concrete subclasses with type information.
```
Comment (jklukas) [inline website/src/contribute/ptransform-style-guide.md]: Changed wording of the last sentence here to remove reference to Java 7 and instead discuss type inferability.

**Comment by jklukas** [comment]
Thanks for the review, @kennknowles.

> But instead of porting tests to InferableFunction can you duplicate them, or some of them.

Done. This should be ready for another review.

**Comment by kennknowles** [comment]
Nice. It looks like the commits should be squashed, but I don't want to assume - is that what you intended?

**Comment by jklukas** [comment]
> It looks like the commits should be squashed

Yes, that would be lovely. Do you need me to do that, or are you happy to use GitHub's squash and merge?

**Comment by kennknowles** [comment]
Done

**Pull request merged**